README.md
<p align="center">
  <a href="https://modx.com/">
    <img alt="MODX Revolution" src="https://raw.githubusercontent.com/modxcms/revolution/2.x/manager/templates/default/images/modx-icon-color.svg" width="100" />
  </a>
</p>
<h1 align="center">
  MODX Revolution
</h1>

#### MODX Revolution is the world’s fastest, most secure, flexible and scalable Open Source CMS

[![LICENSE](https://img.shields.io/badge/License-GPL%20v2-blue.svg)](./LICENSE) [![Build Status](https://github.com/modxcms/revolution/workflows/CI/badge.svg?branch=3.x)](https://github.com/modxcms/revolution/actions?query=branch%3A3.x) [![Contributors](https://img.shields.io/github/contributors/modxcms/revolution.svg)](https://github.com/modxcms/revolution/graphs/contributors) [![Slack Chat](https://img.shields.io/badge/chat_in_slack-online-green.svg?longCache=true&style=flat&logo=slack)](https://modx.org) [![follow on Twitter](https://img.shields.io/twitter/follow/modx.svg?style=social&logo=twitter)](https://twitter.com/intent/follow?screen_name=modx)

## Content Management System and Application Framework

MODX lets you power anything from multi-language, multi-domain corporate sites to personal blogs to mobile APIs. Delivering true creative freedom and removing all restrictions, it lets you control the markup and design without having to code. You can also tailor its modular and extensible core to accommodate virtually any custom requirement or amount of traffic. MODX is the free open source software that meets your needs today—and tomorrow.

MODX Revolution is the world’s fastest, most customizable Open Source PHP CMS. Your creative vision, no restrictions, no compromise.

### Latest Changes

For details read the [complete changelog](./core/docs/changelog.txt 'complete changelog')

### Getting Started

Here's what you need to get started installing or upgrading MODX Revoluton:

+ [Download MODX Revolution](https://modx.com/download)
+ [What's required to run MODX Revolution](https://docs.modx.com/3.x/en/getting-started/server-requirements)
+ [How to install MODX Revolution](https://docs.modx.com/3.x/en/getting-started/installation)
+ [Install MODX via git](https://docs.modx.com/3.x/en/getting-started/installation/git)
+ [How to upgrade from 2.x to 3.0](https://docs.modx.com/3.x/en/getting-started/upgrading-to-3.0)
+ [Read the MODX Documentation](https://docs.modx.com/3.x/en/index)

### Other Important Stuff

MODX is only as good as it is because of many individual community members and users that take the time to [report issues and request new features](https://github.com/modxcms/revolution/issues 'MODX Github Issues'). Make sure you [read the documentation](https://docs.modx.com/3.x/en/index), [post feedback and share your successes](https://community.modx.com/ 'MODX Community') in the MODX community forums. And also help with translating lexicons on [Crowdin platform](https://crowdin.com/project/modx-revolution). To join us in the quest for _Creative Freedom_, [become a Contributor](https://docs.modx.com/3.x/en/contribute/code). You can [contribute using GitHub](https://docs.modx.com/3.x/en/contribute/code 'Contribute to MODX via GitHub')

On behalf of the entire MODX Team,

Thank you for using MODX!

### Security Issues in MODX

For details read the [SECURITY](./SECURITY.md 'SECURITY')

### Copyright

Copyright (c) MODX, LLC. All Rights Reserved.

For complete copyright information, see the [COPYRIGHT](./COPYRIGHT 'Copyright') file at the top-level directory of this distribution.

### License

MODX Revolution is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

[GNU General Public License v2](./LICENSE 'GNU General Public License v2')


changelog.txt
This file shows the changes in this release of xPDO.

xPDO 2.8.5-pl (January 12, 2023)
====================================
- Force query operator to the uppercase (#244)
- Add public xpdo property to xPDOCriteria (#240)

xPDO 2.8.4-pl (February 10, 2022)
====================================
- Fix param typehint for xPDOCache add/set/replace methods (#210)
- Improve PHP 8.1 compatibility (#223)

xPDO 2.8.3-pl (March 31, 2021)
====================================
- Prevent sleep()/benchmark() usage in criteria
- Refactor xPDO->sanitizePKCriteria to use prepared statements
- Prevent unexpected results with non-conditional string criteria

xPDO 2.8.2-pl (February 19, 2021)
====================================
- Update pclzip.lib.php to resolve PHP Notice
- Fix error caused by merging PDO driver options

xPDO 2.8.1-pl (July 13, 2020)
====================================
- Add current_timestamp() to valid currentTimestamps list
- Add float phptype to quotable list to prevent quoting floats as integer

xPDO 2.8.0-pl (February 27, 2020)
====================================
- Load related object attributes correctly into nested relations
- Add option to restrict extraction of files by extension to xPDOZip
- Fix invalid datetime value being used when formatting with regex
- Fix xPDOFileCache->delete() when a directory with same cache key exists
- Fix PHP 7.4 deprecation warnings for array and string offset access with curly braces

xPDO 2.7.0-pl (May 2, 2018)
====================================
- Fix xPDOIterator failure when fetch returns null due to object hydration
- Make json and array phptypes quotable in xPDOQuery
- Replace all deprecated while/each() loops with foreach() loops for PHP 7.2 compat

xPDO 2.6.0-pl (November 1, 2017)
====================================
- Generate engine into the map files for MySQL
- Make InnoDB the default engine for MySQL
- Add missing return statements to built-in validation classes
- Remove all embedded escape characters when escape() is used on a string
- Make sure empty cache file does not return true in PHP format
- Make sure FileCache uses file/folder permissions options + add optional overrides

xPDO 2.5.3-pl (January 3, 2017)
====================================
- Make sure xPDOQuery base class is always loaded

xPDO 2.5.2-pl (December 16, 2016)
====================================
- Sanitize and prevent SQLi in getObject calls expecting PK values
- Make isValidClause a public static method in xPDOQuery
- Allow empty sort direction
- Revert getCriteria change to force scalar params to be PK values
- Make xPDOQuery::isValidClause a public method
- Change constructor in PclZip to be PHP 7 compatible

xPDO 2.5.1-pl (November 14, 2016)
====================================
- Force scalar parameters to be PK values in getCriteria
- Prevent SQLi in sortby, sort direction, and limit clauses
- Detect SQLi clauses when no space exists after UNION

xPDO 2.5.0-pl (February 3, 2016)
====================================
- Optimize and improve logging
- Fix undefined index notice for model indexes
- Allow definition of primary key fields without legacy index="pk"
- Add ARRAY_EXTENDED logTarget that stores an array of data for each log entry

xPDO 2.4.1-pl (September 22, 2015)
====================================
- Optimize xPDO::getOption() for significant performance increase

xPDO 2.4.0-pl (June 26, 2015)
====================================
- Add ability to override mysql table engine via options
- Allow custom attributes in schema object elements
- Allow silent failure of xPDOVehicle validators
- Only show parseBindings results if debug set
- Accept benign values that contain the word union
- Allow xPDOObject::remove() to remove dependents with composite relationships
- Respect engine specified in mysql schema

xPDO 2.3.0-pl (July 15, 2014)
====================================
- Allow recursive xPDOObject::remove() beyond one level
- Allow schema to define callbacks for field
- Remove deprecated xPDO::getMicroTime() method
- Reject xPDOQuery conditions with SQL injection attempts
- Resolve xPDO::parseBindings() error from /e modifier fix
- Fix xPDO::getCount() to work with queries containing group by
- Replace reflection of PHP_SELF with SCRIPT_NAME in xPDO::log()
- Replace use of deprecated /e preg_replace modifier
- Add dependency metadata/API to xPDOTransport

xPDO 2.2.9-pl (September 12, 2013)
====================================
- Improve accuracy of query statistics
- [#10183] Fix cache entry removal on xPDOObject::remove()

xPDO 2.2.8-pl (August 26, 2013)
====================================
- [#9864] Identify compound PK's properly in xPDOQuery::parseConditions()

xPDO 2.2.7-pl (April 9, 2013)
====================================
- [#9793] Preserve numeric keys in xPDOObject::getMany()

xPDO 2.2.6-pl (March 16, 2013)
====================================
- [#7452] Fix sql errors with decimal fields in some locales
- [#9062] Fix problem loading multiple xPDOQuery driver classes
- [#9629] Fix notice in xPDOConnection::getOption() when $options not an array
- [#9020] Prevent xPDOZip from attempting to add empty targets
- [#9272] Implement NFS-safe flock-independent file locking
- [#9020] Allow xPDOZip to pack empty directories
- [#9023] Fix removal of subdirs when deleteTop = false in xPDOCacheManager::deleteTree()
- [#8962] Prevent xPDOSimpleObject from overwriting xPDOObject map entry
- [#8423] Fix tmpLockFile naming issue in xPDOCacheManager::lockFile()
- [#8440] Add memcached implementation of xPDOCache (xPDOMemCached)
- [#8423] Add flock-independent file locking to xPDOCacheManager::writeFile()
- [#6930] Fix improper escaping of $ in parseBindings
- [#8148] Prevent update from criteria using raw NATIVE_KEY on install
- [#8320] Preserve order of $columns in xPDOObject::getSelectColumns()
- Deprecate xPDO->getMicroTime()
- [#8176] xPDO->updateCollection() fails to set NULL values

xPDO 2.2.5-pl (June 1, 2012)
====================================

pdo.class.php
<?php
/*
 * OpenExpedio (xPDO)
 * Copyright (C) 2006 Jason Coward <xpdo@opengeek.com>
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 330, Boston, MA 02111-1307 USA
 */
 
/**
 * The main class for the PDO for PHP 4 implementation.
 * 
 * Provides limited PHP 4 compatible {@link PDO} and {@link PDOStatement}
 * implementations for PHP 4 (and 5.0.x platforms without the PECL PDO
 * extensions installed).
 * 
 * @package xpdo
 */

/**
 * The PHP Data Objects (PDO) extension defines a lightweight, consistent
 * interface for accessing databases in PHP. Each database driver that
 * implements the PDO interface can expose database-specific features as regular
 * extension functions. Note that you cannot perform any database functions
 * using the PDO extension by itself; you must use a database-specific PDO
 * driver to access a database server.  See the {@link http://php.net/xpdo
 * official PDO documentation} for more detailed information on PDO and how it
 * is used.
 * 
 * This PDO implementation is provided by OpenExpedio (XPDO) to allow
 * compatibility with PHP 4.3+ when native PDO is not available (PHP 5.1+).
 * 
 * @package xpdo
 * @see http://php.net/xpdo
 * @todo Implement additional driver support; currently only supports MySQL,
 * PostgreSQL, and SQLite
 */
class PDO {
    var $_driver;
    var $_dbtype= null;

    /**
     * Returns a list of available database drivers.
     * 
     * @return array All drivers available to any PDO instance.
     * @todo Implement this as a dynamic list of available drivers.
     */
    function getAvailableDrivers() {
        return array (
            'mysql',
            'pgsql',
            'sqlite'
        );
    }

    /**
     * Creates a PDO instance to represent a connection to the requested
     * database. This is the public constructor for PHP 4 support; falls through
     * to PHP 5 style constructor, and is ignored by PHP 5.
     * 
     * @see __construct
     */
    function PDO($string_dsn, $string_username= '', $string_password= '', $array_driver_options= null) {
        $this->__construct($string_dsn, $string_username, $string_password, $array_driver_options);
    }
    /**
     * Creates a PDO instance to represent a connection to the requested database.
     * 
     * @see http://php.net/manual/en/function.pdo-construct.php
     * @param string $dsn The Data Source Name, or DSN, contains the information
     * required to connect to the database.
     * @param string $username The user name for the DSN string. This parameter
     * is optional for some PDO drivers.
     * @param string $password The password for the DSN string. This parameter
     * is optional for some PDO drivers.
     * @param array $driver_options A key=>value array of driver-specific
     * connection options.
     * @return A valid PDO instance if successful; error/exception is raised
     * otherwise.
     */
    function __construct($string_dsn, $string_username= '', $string_password= '', $array_driver_options= null) {
        $con= xPDO :: parseDSN($string_dsn);
        $driverClass= XPDO_CORE_PATH . 'pdo.' . $con['dbtype'] . '.inc.php';
        include_once (XPDO_CORE_PATH . 'pdo.' . $con['dbtype'] . '.inc.php');
        $this->_dbtype= $con['dbtype'];
        if ($con['dbtype'] === 'mysql') {
            if (isset ($con['port']))
                $con['host'] .= ':' . $con['port'];
            $this->_driver= new PDO_mysql($con['host'], $con['dbname'], $string_username, $string_password);
        }
        elseif ($con['dbtype'] === 'sqlite2' || $con['dbtype'] === 'sqlite') {
            $this->_driver= new PDO_sqlite($con['dbname']);
        }
        elseif ($con['dbtype'] === 'pgsql') {
            $string_dsn= 'host=' . $con['host'] . ' dbname=' . $con['dbname'] . ' user=' . $string_username . ' password=' . $string_password;
            if (isset ($con['port']))
                $string_dsn .= ' port=' . $con['port'];
            $this->_driver= new PDO_pgsql($string_dsn);
        }
    }

    /** 
     * @see http://php.net/manual/en/function.pdo-begintransaction.php
     */
    function beginTransaction() {
        $this->_driver->beginTransaction();
    }

    /** 
     * @see http://php.net/manual/en/function.pdo-commit.php
     */
    function commit() {
        $this->_driver->commit();
    }

    /** 
     * @see http://php.net/manual/en/function.pdo-exec.php
     */
    function exec($query) {
        return $this->_driver->exec($query);
    }

    /** 
     * @see http://php.net/manual/en/function.pdo-errorcode.php
     */
    function errorCode() {
        return $this->_driver->errorCode();
    }

    /** 
     * @see http://php.net/manual/en/function.pdo-errorinfo.php
     */
    function errorInfo() {
        return $this->_driver->errorInfo();
    }

    /** 
     * @see http://php.net/manual/en/function.pdo-getattribute.php
     */
    function getAttribute($attribute) {
        return $this->_driver->getAttribute($attribute);
    }

    /** 
     * @see http://php.net/manual/en/function.pdo-lastinsertid.php
     */
    function lastInsertId() {
        return $this->_driver->lastInsertId();
    }

    /** 
     * @see http://php.net/manual/en/function.pdo-prepare.php
     */
    function prepare($statement, $driver_options= array ()) {
        return $this->_driver->prepare($statement, $driver_options= array ());
    }

    /** 
     * @see http://php.net/manual/en/function.pdo-query.php
     */
    function query($query) {
        return $this->_driver->query($query);
    }

    /** 
     * @see http://php.net/manual/en/function.pdo-quote.php
     */
    function quote($string, $parameter_type= PDO_PARAM_STR) {
        return $this->_driver->quote($string, $parameter_type);
    }

    /** 
     * @see http://php.net/manual/en/function.pdo-rollback.php
     */
    function rollBack() {
        $this->_driver->rollBack();
    }

    /** 
     * @see http://php.net/manual/en/function.pdo-setattribute.php
     */
    function setAttribute($attribute, $value) {
        return $this->_driver->setAttribute($attribute, $value);
    }
}

/**
 * Represents a PDO prepared statement.
 */
class PDOStatement {
    /**
     * @var string The SQL query string for the statement to use.
     */
    var $queryString= '';

    var $_connection;
    var $_dbinfo;
    var $_persistent= false;
    var $_result= null;
    var $_fetchmode= PDO_FETCH_BOTH;
    var $_errorCode= '';
    var $_errorInfo= array (
        PDO_ERR_NONE
    );
    var $_boundParams= array ();
    
    function PDOStatement($queryString, & $connection, & $dbinfo) {
        $this->__construct($queryString, $connection, $dbinfo);
    }
    function __construct($queryString, & $connection, & $dbinfo) {
        $this->queryString= $queryString;
        $this->_connection= & $connection;
        $this->_dbinfo= & $dbinfo;
    }

    function bindParam($param, & $variable, $data_type= PDO_PARAM_STR, $length= 0, $driver_options= null) {
        $this->_boundParams[$param]['value']= $variable;
        $this->_boundParams[$param]['type']= $data_type;
        $this->_boundParams[$param]['length']= intval($length);
    }

    function bindValue($param, $value, $data_type= PDO_PARAM_STR) {
        $this->_boundParams[$param]['value']= $value;
        $this->_boundParams[$param]['type']= $data_type;
        $this->_boundParams[$param]['length']= 0;
    }

    function errorCode() {
        return $this->_errorCode;
    }

    function errorInfo() {
        return $this->_errorInfo;
    }

    function execute($input_parameters= null) {
        $array= & $this->_boundParams;
        if (is_array($input_parameters) && !empty ($input_parameters)) {
            $array= $input_parameters;
        }
        $queryString= $this->queryString;
        if (count($array) > 0) {
            reset($array);
            while (list ($k, $param)= each($array)) {
                $v= $param['value'];
                $type= $param['type'];
                if (!is_int($k) || substr($k, 0, 1) === ':') {
                    if (!isset ($tempf)) {
                        $tempf= $tempr= array ();
                    }
                    $pattern= '/' . $k . '\b/';
                    array_push($tempf, $pattern);
                    $v= $this->quote($v, $type);
                    array_push($tempr, $v);
                } else {
                    $parse= create_function('$d,$v,$t', 'return $d->quote($v, $t);');
                    $queryString= preg_replace("/(\?)/e", '$parse($this,$array[$k],$type);', $queryString, 1);
                }
            }
            if (isset ($tempf)) {
                $queryString= preg_replace($tempf, $tempr, $queryString);
            }
        }
        if (is_null($this->_result= $this->_uquery($queryString))) {
            $keyvars= false;
        } else {
            $keyvars= true;
        }
        return $keyvars;
    }

    function setFetchMode($mode) {
        $result= false;
        if ($mode= intval($mode)) {
            switch ($mode) {
                case PDO_FETCH_NUM :
                case PDO_FETCH_ASSOC :
                case PDO_FETCH_OBJ :
                case PDO_FETCH_BOTH :
                default:
                    $result= true;
                    $this->_fetchmode= $mode;
                    break;
            }
        }
        return $result;
    }

    function closeCursor() {
        do {
           while ($this->fetch()) {}
           if (!$this->nextRowset())
               break;
        } while (true);
    }

    function bindColumn($column, & $param, $type= null, $max_length= null, $driver_option= null) { return false; }

    function columnCount() { return false; }

    function getAttribute($attribute) { return false; }

    function getColumnMeta($column) { return false; }

    function fetch($mode= PDO_FETCH_BOTH, $cursor= null, $offset= null) { return false; }

    function fetchAll($mode= PDO_FETCH_BOTH, $column_index= 0) { return false; }

    function fetchColumn($column_number= 0) { return false; }

    function fetchObject($class_name= '', $ctor_args= null) { return false; }
    
    function nextRowset() { return false; }

    function quote($string, $parameter_type= PDO_PARAM_STR) { return false; }
    
    function rowCount() { return false; }

    function setAttribute($attribute, $value) { return false; }
    
    function debugDumpParams() { return false; }

    function _setErrors($er) { return false; }

    function _uquery(& $query) { return false; }
}
?>
